File: 0005 Longest Palindromic Substring/solution.py
#!python3

class Solution:

    def longestPalindrome(self, s):
        n = len(s)
        if n <= 1:
            return s
        
        dp = [[False for _ in range(n)] for _ in range(n)]
        
        res = s[0]
        for r in range(1, n):
            for l in range(r):
                if s[l] == s[r] and (r - l <= 2 or dp[l + 1][r - 1]):
                    dp[l][r] = True
                    if (r - l + 1) > len(res):
                        res = s[l: r + 1]
        return res

if __name__ == "__main__":
    s = Solution()
    print(s.longestPalindrome("abba") == "abba")
    print(Solution().longestPalindrome("babad") == "bab")
    print(Solution().longestPalindrome("cbbd") == "bb")

File: 0005 Longest Palindromic Substring/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) 
    {
        string res;
        for (int i = 0; i < s.size(); i++)
        {
            //最长回文子串长度为奇数
            for (int j = i, k = i; j >= 0 && k < s.size() && s[j] == s[k]; j--, k++)
                if (res.size() < k - j + 1)
                    res = s.substr(j, k - j + 1);
            
            //最长回文子串长度为偶数
            for (int j = i, k = i + 1; j >= 0 && k < s.size() && s[j] == s[k]; j--, k++)
                if (res.size() < k - j + 1)
                    res = s.substr(j, k - j + 1);
        }
        
        return res;
    }
};

int main(){
    Solution s;
    assert(s.longestPalindrome("babad") == "bab");
    assert(s.longestPalindrome("cbbd") == "bb");
    return 0;
}